core/src/main/java/org/apache/accumulo/core/client/impl/Tables.java

```diff
@@ -129,12 +86,26 @@ public static String getTableName(Instance instance, String tableId) throws Tabl
     return tableName;
   }
 
-  public static SortedMap<String,String> getNameToIdMap(Instance instance) {
-    return getMap(instance, true);
+  public static Map<String,String> getNameToIdMap(Instance instance) {
+    TableMap map = tableMapCache;
+    // check if map needs to be updated
+    if (map == null || tableMapTimestamp.longValue() + TABLE_MAP_CACHE_EXPIRATION <= System.nanoTime()) {
+      tableMapCache = new TableMap(instance);
+      tableMapTimestamp.set(System.nanoTime());
+      map = tableMapCache;
+    }
+    return map.getNameToIdMap();
   }
 
-  public static SortedMap<String,String> getIdToNameMap(Instance instance) {
-    return getMap(instance, false);
+  public static Map<String,String> getIdToNameMap(Instance instance) {
+    TableMap map = tableMapCache;
+    // check if map needs to be updated
+    if (map == null || tableMapTimestamp.longValue() + TABLE_MAP_CACHE_EXPIRATION <= System.nanoTime()) {
```

Comment:
Could make the TableMap class have a final long creationTime field instead of having the mutable tableMapTimestamp.  Then the CPU may be able to avoid two trips to main memory.  Also I think keeping it all together and immutable makes it easier to reason about.

Could also do a double check to avoid multiple threads regenerating and put all of this in a single function.

```java
private static TableMap getTableMap() {
  TableMap map = tableMapCache;
  if (map == null || map.getCreationTime() + TABLE_MAP_CACHE_EXPIRATION <= System.nanoTime()) {
    synchronized(Tables.class) {
       map = tableMapCache;
       if (map == null || map.getCreationTime() + TABLE_MAP_CACHE_EXPIRATION <= System.nanoTime()) {
         map = tableMapCache = new TableMap(instance);  //this could set creation time, should do that last in constructor
       }     
    }
  }
  return map;
}
```

Reply:
I had thought about this as well but wasn't sure where to put it in TableMap.  You have this replacing the constructor?

Reply:
I was thinking of a new static method in Tables class